Use our own instance of `PROBLEME_ANTARES_A_RESOUDRE`

flomnes opened the pull request:
- Pass `ProblemeHebdo` by const pointer
- Pass HourlyCSRProblem's `PROBLEME_ANTARES_A_RESOUDRE` by pointer. Use a different instance than the member of `PROBLEME_HEBDO` since it has nothing in common (weekly adequacy problem _vs_ hourly CSR problem)
- Use anonymous namespace according to @sylvlecl suggestion (avoid exposing implementation details, better encapsulation)

### TODO
- [X] `HourlyCSRProblem::setVariableBounds()`
- [X] `HourlyCSRProblem::setProblemCost()`
- [x] `HourlyCSRProblem::buildProblemConstraintsRHS()`
- [x] `HourlyCSRProblem::buildProblemConstraintsLHS()`
- [x] Handle memory allocation/deallocation
- [x] Invesitate `OPT_LiberationProblemesSimplexe(problemeHebdo_);` : not sure we want to call this

flomnes committed:
Simplify UtiliserLaToleranceDAdmissibiliteParDefaut, etc.

flomnes committed:
Use PROBLEME_POINT_INTERIEUR for calculateCsrCostFunctionValue

flomnes committed:
Pass hour instead of HOURLY_CSR_PROBLEM&

flomnes committed:
Use PROBLEME_POINT_INTERIEUR for CSR_DEBUG_HANDLE, handleInteriorPointError

flomnes committed:
PR remarks (use ref on object, not unique_ptr)

flomnes committed:
Rename HOURLY_CSR_PROBLEM -> HourlyCSRProblem

flomnes committed:
Apply clang-format

flomnes committed:
Fix conflicts with base branch

flomnes committed:
Move files

flomnes committed:
Move csr_quadratic_problem.h,cpp

flomnes committed:
Remove unused fwd declaration

flomnes committed:
Don't instantiate HourlyCSRProblem for every hour, setHour instead

flomnes committed:
Rename OPT_ConstruireLaMatriceDesContraintesDuProblemeQuadratique_CSR

flomnes committed:
Use std::vector::data instead of malloc/free

flomnes committed:
Merge remote-tracking branch 'github/feature/adq-patch-final' into fix/csr-quadratic-problem

flomnes committed:
Some renaming

flomnes committed:
Use our own instance of PROBLEME_ANTARES_A_RESOUDRE

guilpier-code commented on src/solver/optimisation/hourly_csr_problem.h:
```diff
@@ -45,6 +45,7 @@ class HourlyCSRProblem
     int hourInWeekTriggeredCsr;
     double belowThisThresholdSetToZero;
     PROBLEME_HEBDO* problemeHebdo_;
+    PROBLEME_ANTARES_A_RESOUDRE problemeAResoudre_;
```
**Question** : 
We have a **problemeHebdo_** as a member of class **HourlyCSRProblem**. To me, it should not be (an weekly problem should not be part of a hourly problem). Is is a transitional stage of the code ? Is it intended to remain in the final state ?

flomnes replied:
Yeah we need it for many things:
- Adequacy patch parameters
- Adequacy patch runtime data (e.g areas inside/outside/virtual)
- Variable/constraint correspondence arrays necessary for
  - Accessing (read & write) results from the adequacy simulation
  - Accessing the hourly binding constraint indices (input data)

It would be nice to remove it though.

guilpier-code commented on src/solver/optimisation/adequacy_patch_csr/solve_problem.cpp:
```diff
@@ -256,20 +256,20 @@ void handleInteriorPointError(const PROBLEME_POINT_INTERIEUR& Probleme,
                    int yearNb)
 {
-    auto Probleme = buildInteriorPointProblem(ProblemeAResoudre);
+    auto problemeAResoudre = &hourlyCsrProblem.problemeAResoudre_;
```
Why not : 
```C++
    auto problemeAResoudre = hourlyCsrProblem->problemeAResoudre_;
```
?

flomnes replied:
Because the function after takes a pointer as an argument. It could take a non-const ref.

flomnes committed:
Refactor 2 more functions

OPT_InitialiserLeSecondMembreDuProblemeQuadratique_CSR
OPT_ConstruireLaListeDesVariablesOptimiseesDuProblemeQuadratique_CSR

flomnes committed:
Merge remote-tracking branch 'github/feature/adq-patch-final' into fix/own-probleme-a-resoudre

flomnes committed:
Use reference instead of raw pointer

flomnes committed:
Extract allocation function for PROBLEME_ANTARES_A_RESOUDRE

sylvlecl commented on src/solver/optimisation/adequacy_patch_csr/set_problem_cost_function.cpp:
```diff
@@ -113,47 +116,50 @@ void setLinearCost(PROBLEME_HEBDO* ProblemeHebdo, int hour)
 
-    PROBLEME_ANTARES_A_RESOUDRE* ProblemeAResoudre = ProblemeHebdo->ProblemeAResoudre;
-    memset((char*)ProblemeAResoudre->CoutLineaire,
+    memset((char*)ProblemeAResoudre.CoutLineaire,
```
I know it's not part of the PR, put this cast to `char*` seems suspect to me ?

flomnes replied:
It does

sylvlecl commented on src/solver/optimisation/adequacy_patch_csr/construct_problem_variables.cpp:
```diff
@@ -65,14 +67,14 @@ void constructVariableENS(PROBLEME_HEBDO* ProblemeHebdo, HourlyCSRProblem& hourl
-    CORRESPONDANCES_DES_VARIABLES* CorrespondanceVarNativesVarOptim
-        = ProblemeHebdo->CorrespondanceVarNativesVarOptim[hour];
-    int& NumberOfVariables = ProblemeAResoudre->NombreDeVariables;
+    CORRESPONDANCES_DES_VARIABLES* CorrespondanceVarNativesVarOptim;
```
It's considered better practice to have declaration and assignment on the same line

flomnes replied:
done

flomnes committed:
Use std::fill_n instead of memset

flomnes committed:
PR remark

flomnes committed:
Extract OPT_FreeOptimizationData

flomnes committed:
Handle memory allocation/deallocation
